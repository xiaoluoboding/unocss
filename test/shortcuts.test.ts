import { createGenerator } from '@unocss/core'
import presetUno from '@unocss/preset-uno'
import prettier from 'prettier/standalone'
import parserCSS from 'prettier/parser-postcss'
import { describe, expect, test } from 'vitest'

describe('shortcuts', () => {
  const uno = createGenerator({
    shortcuts: [
      {
        sh1: 'px-2 py-3 sh3',
        sh2: 'hover:text-lg text-sm text-lg',
        sh3: 'sm:m-2 m-3',
      },
      {
        btn: 'mr-10',
        btn1: 'ml-10 btn',
        btn2: 'sh1 btn1',
      },
      [/^button-(\d)$/, ([, d]) => [`px${(+d) * 3}`, `py${(+d) * 2}`]],
      ['bad-one', 'p2 unmatched'],
      ['transform-duplicated', 'translate-x-1 translate-y-2 scale-4 hover:scale-2 active:scale-x-4'],
    ],
    presets: [
      presetUno(),
    ],
  })

  test('static', async() => {
    const { css } = await uno.generate('sh1 sh2 focus:sh2 sh3')
    expect(css).toMatchSnapshot()
  })

  test('nesting static', async() => {
    const { css } = await uno.generate('btn1 btn btn2')
    expect(css).toMatchSnapshot()
  })

  test('dynamic', async() => {
    const { css } = await uno.generate('button-1 button-2')
    expect(css).toMatchSnapshot()
  })

  test('merge transform-duplicated', async() => {
    const { css } = await uno.generate('transform-duplicated')
    const prettified = prettier.format(css, {
      parser: 'css',
      plugins: [parserCSS],
    })
    expect(prettified).toMatchSnapshot()
  })
})
