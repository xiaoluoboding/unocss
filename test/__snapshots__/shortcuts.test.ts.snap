// Vitest Snapshot v1

exports[`shortcuts > dynamic 1`] = `
"/* layer: shortcuts */
.button-1{padding-left:0.75rem;padding-right:0.75rem;padding-top:0.5rem;padding-bottom:0.5rem;}
.button-2{padding-left:1.5rem;padding-right:1.5rem;padding-top:1rem;padding-bottom:1rem;}"
`;

exports[`shortcuts > merge transform-duplicated 1`] = `
"/* layer: shortcuts */
.transform-duplicated:hover {
  --un-scale-x: 0.02;
  --un-scale-y: 0.02;
}
.transform-duplicated {
  --un-rotate: 0;
  --un-scale-x: 1;
  --un-scale-y: 1;
  --un-scale-z: 1;
  --un-skew-x: 0;
  --un-skew-y: 0;
  --un-translate-x: 0;
  --un-translate-y: 0;
  --un-translate-z: 0;
  transform: rotate(var(--un-rotate)) scaleX(var(--un-scale-x))
    scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z)) skewX(var(--un-skew-x))
    skewY(var(--un-skew-y)) translateX(var(--un-translate-x))
    translateY(var(--un-translate-y)) translateZ(var(--un-translate-z));
  --un-translate-x: 0.25rem;
  --un-translate-y: 0.5rem;
  --un-scale-x: 0.04;
  --un-scale-y: 0.04;
}
.transform-duplicated:active {
  --un-scale-x: 0.04;
}
"
`;

exports[`shortcuts > nesting static 1`] = `
"/* layer: shortcuts */
.btn{margin-right:2.5rem;}
.btn1{margin-left:2.5rem;margin-right:2.5rem;}
.btn2{margin:0.75rem;margin-left:2.5rem;margin-right:2.5rem;padding-left:0.5rem;padding-right:0.5rem;padding-top:0.75rem;padding-bottom:0.75rem;}
@media (min-width: 640px){
.btn2{margin:0.5rem;}
}"
`;

exports[`shortcuts > static 1`] = `
"/* layer: shortcuts */
.sh3{margin:0.75rem;}
.sh1{margin:0.75rem;padding-left:0.5rem;padding-right:0.5rem;padding-top:0.75rem;padding-bottom:0.75rem;}
.focus\\\\:sh2:focus,.sh2{font-size:0.875rem;line-height:1.25rem;font-size:1.125rem;line-height:1.75rem;}
.focus\\\\:sh2:hover:focus,.sh2:hover{font-size:1.125rem;line-height:1.75rem;}
@media (min-width: 640px){
.sh1,.sh3{margin:0.5rem;}
}"
`;
